v2/api_error_test.go

```diff
@@ -0,0 +1,169 @@
+	}
+	st, _ := status.New(codes.FailedPrecondition, "pf").WithDetails(pf)
+	apierr := &APIError{
+		err:     st.Err(),
+		status:  st,
+		details: ErrDetails{PreconditionFailure: pf},
+	}
+	if !strings.Contains(apierr.Error(), "Foo") {
+		t.Errorf("Status message not included!")
+	}
+}
+
+func TestGRPCStatus(t *testing.T) {
+	qf := &errdetails.QuotaFailure{
+		Violations: []*errdetails.QuotaFailure_Violation{{Subject: "Foo", Description: "Bar"}},
+	}
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	apierr := &APIError{
+		err:     st.Err(),
+		status:  st,
+		details: ErrDetails{QuotaFailure: qf},
+	}
+	if apierr.status != apierr.GRPCStatus() {
+		t.Errorf("Expected: %v but got: %v", apierr.status, apierr.GRPCStatus())
+	}
+}
+
+func TestFromError(t *testing.T) {
+	m := make(map[string]string)
+	m["type"] = "ErrorInfo"
+	ei := &errdetails.ErrorInfo{
+		Reason:   "Foo",
+		Domain:   "Bar",
+		Metadata: m,
+	}
+	eS, _ := status.New(codes.Unauthenticated, "ei").WithDetails(ei)
+
+	br := &errdetails.BadRequest{
+		FieldViolations: []*errdetails.BadRequest_FieldViolation{{
+			Field:       "Foo",
+			Description: "Bar",
+		}},
+	}
+	bS, _ := status.New(codes.InvalidArgument, "br").WithDetails(br)
+
+	qf := &errdetails.QuotaFailure{
+		Violations: []*errdetails.QuotaFailure_Violation{{Subject: "Foo", Description: "Bar"}},
+	}
+	qS, _ := status.New(codes.ResourceExhausted, "qf").WithDetails(qf, br)
+
+	pf := &errdetails.PreconditionFailure{
+		Violations: []*errdetails.PreconditionFailure_Violation{{Type: "Foo", Subject: "Bar", Description: "desc"}},
+	}
+	pS, _ := status.New(codes.FailedPrecondition, "pf").WithDetails(pf)
+
+	ri := &errdetails.RetryInfo{
+		RetryDelay: &durationpb.Duration{Seconds: 10, Nanos: 10},
+	}
+	riS, _ := status.New(codes.Unavailable, "foo").WithDetails(ri)
+
+	rs := &errdetails.ResourceInfo{
+		ResourceType: "Foo",
+		ResourceName: "Bar",
+		Owner:        "Client",
+		Description:  "Directory not Found",
+	}
+	rS, _ := status.New(codes.NotFound, "rs").WithDetails(rs)
+
+	rq := &errdetails.RequestInfo{
+		RequestId:   "Foo",
+		ServingData: "Bar",
+	}
+	rqS, _ := status.New(codes.Canceled, "Request cancelled by client").WithDetails(rq)
+
+	deb := &errdetails.DebugInfo{
+		StackEntries: []string{"Foo", "Bar"},
+		Detail:       "Stack",
+	}
+	dS, _ := status.New(codes.DataLoss, "Here is the debug info").WithDetails(deb)
+
+	hp := &errdetails.Help{
+		Links: []*errdetails.Help_Link{{Description: "Foo", Url: "Bar"}},
+	}
+	hS, _ := status.New(codes.Unimplemented, "Help Info").WithDetails(hp)
+
+	lo := &errdetails.LocalizedMessage{
+		Locale:  "Foo",
+		Message: "Bar",
+	}
+	lS, _ := status.New(codes.Unknown, "Localized Message").WithDetails(lo)
+
+	tests := []struct {
+		apierr *APIError
+		b      bool
+	}{
+		{&APIError{err: eS.Err(), status: eS, details: ErrDetails{ErrorInfo: ei}}, true},
```

Comment:
Great, great job taking the feedback I had before and improving this code. I challenge you to optimize it further. Think about what parts of the test matter. Does the `Status.Code` actually matter for the test? Could it be anything? How about the `Status.Message`? Neither really affects the execution of `APIError` code. The only input that really matters is the Details. We could move the `status.New` and the `&APIError{}` construction into the `for` loop.

Reply:
Thank you so much! For the status.New, I had tried including it in the for loop but the WithDetails() method was an issue because it returns back two outputs(status, error). When I included status in the test struct, I was having a difficult time thinking of how to initialize both the status and error message that would be returned from WithDetails() which was why I did it outside the for loop but I'll take another look and see if there's a way around